kvbc/src/sparse_merkle/proof_path_processor.cpp

```diff
@@ -0,0 +1,146 @@
+// Concord
+//
+// Copyright (c) 2023 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the
+// LICENSE file.
+
+#include "sparse_merkle/proof_path_processor.h"
+#include "sparse_merkle/internal_node.h"
+#include "sparse_merkle/keys.h"
+
+namespace concord {
+namespace kvbc {
+namespace sparse_merkle {
+namespace proof_path_processor {
+
+using Sliver = concordUtils::Sliver;
+
+bool verifyProofPath(Sliver key, Sliver value, const std::vector<Hash>& proofPath, const Hash& rootHash) {
+  Hasher hasher{};
+  auto keyHash = hasher.hash(key.data(), key.length());
+  auto valueHash = hasher.hash(value.data(), value.length());
+
+  enum class Direction : std::uint8_t { Left, Right };
+  const auto BatchedInternalNodeLevelsIterated = BatchedInternalNode::MAX_HEIGHT - 1;
+
+  std::vector<Direction> pathOrdering;
+  for (int nibble_index = 0; nibble_index < Hash::MAX_NIBBLES; nibble_index++) {
+    auto current_nibble = keyHash.getNibble(nibble_index).data();
+    auto node_index = BatchedInternalNode::nibbleToIndex(current_nibble);
+    std::vector<Direction> subPath;
```

Comment:
Can be made an array and save up the allocation

Reply:
Yes, here we will benefit from using an array and the additional complexity will remain hidden. Will make the change.